File: go/domain/domain.go
package domain

type BranchCreated struct {
	Name    string
	Country string
}

func (b BranchCreated) IsEventPayload() bool {
	return true
}

type BranchNameChanged struct {
	Name string
}

func (b BranchNameChanged) IsEventPayload() bool {
	return true
}

type Branch struct {
	Name    string
	Country string
}

func (b Branch) IsAggregatePayload() bool {
	return true
}

type BranchProjector struct{}

func (p BranchProjector) GetVersion() string {
	return "1.0"
}

func (p BranchProjector) Project(payload AggregatePayload, ev *EventCommon) AggregatePayload {
	switch ap := payload.(type) {
	case EmptyAggregatePayload:
		switch ep := ev.Payload.(type) {
		case BranchCreated:
			return Branch{Name: ep.Name, Country: ep.Country}
		default:
			return payload
		}
	case Branch:
		switch ep := ev.Payload.(type) {
		case BranchNameChanged:
			return Branch{
				Name:    ep.Name,
				Country: ap.Country, // 他のフィールドは元の値をコピー
			}
		default:
			return payload
		}
	default:
		return payload
	}
}

type CreateBranchCommand struct {
	Name    string
	Country string
}

func (c CreateBranchCommand) IsCommand() bool {
	return true
}

type ChangeBranchNameCommand struct {
	Name          string
	PartitionKeys PartitionKeys
}

func (c ChangeBranchNameCommand) IsCommand() bool {
	return true
}

File: typescript/src/domain.ts
import { AggregatePayload, AggregateProjector, EmptyAggregatePayload, EventCommon, EventPayload, Command, CommandWithHandler, CommandContext, EventPayloadOrNone, PartitionKeys } from "./lib";

export class BranchCreated implements EventPayload {
    type: string = 'BranchCreated';
    constructor(public readonly name: string, public readonly country: string) {}
    IsEventPayload(): boolean {
        return true;
    }
}
export class BranchNameChanged implements EventPayload {
    type: string = 'BranchNameChanged';
    constructor(public readonly name: string) {}
    IsEventPayload(): boolean {
        return true;
    }
}
export class BranchCountryChanged implements EventPayload {
    type: string = 'BranchCountryChanged';
    constructor(public country: string) {}
    IsEventPayload(): boolean {
        return true;
    }
}

export class Branch implements AggregatePayload {
    type: string = 'Branch';
    constructor(public name: string, public country: string) {}
    IsAggregatePayload() : boolean {
        return true;
    }
}

export class BranchProjector implements AggregateProjector {
    constructor() {}
    Project(payload: AggregatePayload, ev: EventCommon): AggregatePayload {
        if (payload instanceof Branch) {
            if (ev.Payload instanceof BranchNameChanged) {
                return new Branch(ev.Payload.name, payload.country);
            } else if (ev.Payload instanceof BranchCountryChanged) {
                return new Branch(payload.name, ev.Payload.country);
            }
        } else if (payload instanceof EmptyAggregatePayload && ev.Payload instanceof BranchCreated) {
            return new Branch(ev.Payload.name, ev.Payload.country);
        }
        return payload;
    }
    GetVersion(): string { return '1.0.0'; }

}

export class CreateBranch implements Command {
    IsCommand(): boolean {
        return true;
    }
    constructor(public readonly name: string, public readonly country: string) {}
}

export class ChangeBranchName implements CommandWithHandler<ChangeBranchName> {

    constructor(public readonly name: string, public readonly partitionKeys: PartitionKeys ) {}
    GetAggregateProjector(): AggregateProjector {
        return new BranchProjector();
    }

    PartitionKeysProvider(command: ChangeBranchName): PartitionKeys {
        return command.partitionKeys;
    }
    CommandHandler(command: ChangeBranchName, context: CommandContext): EventPayloadOrNone {
        return new BranchNameChanged(command.name);
    }
    IsCommand(): boolean {
        return true;
    }

}